lib/link_checker.rb

```diff
@@ -0,0 +1,75 @@
+class LinkChecker
+  class << self
+    def for_files_at(path)
+      path = File.join(path, "**/*.erb") unless path.to_s.include?("*")
+      texts = Dir.glob(path).map { |filename| File.read(filename) }
+      new(texts)
+    end
+  end
+
+  attr_reader :texts
+
+  def initialize(texts)
+    @texts = texts
+  end
+
+  def urls
+    @urls ||= extract_urls
+  end
+
+  def report_text
+    content = []
+    content << "LINKS TO CHECK #{urls.length}"
+    content << "Broken links #{report.totals.broken}"
+    content << "Cautious links #{report.totals.caution}"
+    report.links.each do |link|
+      next unless link.status == :broken || link.status == :caution
+
+      content << "#{link.uri} - #{link.problem_summary}"
+    end
+    content.join("\n")
+  end
+
+  def report
+    @report ||= generate_report
+  end
+
+private
+
+  def extract_urls
+    return Set[] if texts.blank?
+
+    urls = texts.map { |text| URI.extract(text, %w[http https]) }
+    urls.flatten!
+    urls.map! { |uri| remove_unmatched_closing_bracket_from_end(uri) }
+    urls.to_set
+  end
+
+  def remove_unmatched_closing_bracket_from_end(text)
+    text.gsub!(/\)\W?$/, "") if unmatched_closing_bracket?(text)
+    text
+  end
+
+  def unmatched_closing_bracket?(text)
+    text.count("(") < text.count(")")
+  end
```

Comment:
Not sure this needs to be separate method? I can't see this being used anywhere else expect once in the `remove_unmatched_closing_bracket_from_end` method

Reply:
I think it makes the code easier to read.